File: shared/senderActiveSchedule.ts
export type SenderScheduleMode = 'blacklist' | 'whitelist'

export type SenderActiveScheduleRange = {
  start: string
  end: string
}

export type SenderActiveScheduleRule = {
  enabled: boolean
  mode: SenderScheduleMode
  weekdays: number[]
  ranges: SenderActiveScheduleRange[]
}

export type SenderActiveSchedule = {
  sms: SenderActiveScheduleRule
  appNotify: SenderActiveScheduleRule
  callNotify: SenderActiveScheduleRule
}

const ALL_WEEKDAYS = [1, 2, 3, 4, 5, 6, 7] as const
const TIME_PATTERN = /^([01]\d|2[0-3]):([0-5]\d)$/

export function buildDefaultSenderActiveScheduleRule(): SenderActiveScheduleRule {
  return {
    enabled: false,
    mode: 'blacklist',
    weekdays: [...ALL_WEEKDAYS],
    ranges: []
  }
}

export function buildStarterSenderActiveScheduleRule(): SenderActiveScheduleRule {
  return {
    enabled: true,
    mode: 'blacklist',
    weekdays: [...ALL_WEEKDAYS],
    ranges: [{ start: '09:00', end: '18:00' }]
  }
}

export function buildDefaultSenderActiveSchedule(): SenderActiveSchedule {
  return {
    sms: buildDefaultSenderActiveScheduleRule(),
    appNotify: buildDefaultSenderActiveScheduleRule(),
    callNotify: buildDefaultSenderActiveScheduleRule()
  }
}

export function normalizeSenderActiveSchedule(value: unknown): SenderActiveSchedule {
  const raw = isPlainObject(value) ? value : {}
  return {
    sms: normalizeSenderActiveScheduleRule(raw.sms),
    appNotify: normalizeSenderActiveScheduleRule(raw.appNotify),
    callNotify: normalizeSenderActiveScheduleRule(raw.callNotify)
  }
}

export function normalizeSenderActiveScheduleRule(value: unknown): SenderActiveScheduleRule {
  const raw = isPlainObject(value) ? value : {}
  const ranges = normalizeRanges(raw.ranges)
  return {
    enabled: raw.enabled === true && ranges.length > 0,
    mode: raw.mode === 'whitelist' ? 'whitelist' : 'blacklist',
    weekdays: normalizeWeekdays(raw.weekdays),
    ranges
  }
}

export function countEnabledScheduleRanges(rule: SenderActiveScheduleRule): number {
  return rule.enabled ? rule.ranges.length : 0
}

function normalizeWeekdays(value: unknown): number[] {
  if (!Array.isArray(value)) return [...ALL_WEEKDAYS]
  const normalized = value
    .map((item) => (typeof item === 'number' ? item : Number(item)))
    .filter((item) => Number.isInteger(item) && item >= 1 && item <= 7)
    .filter((item, index, array) => array.indexOf(item) === index)
  return normalized.length ? normalized : [...ALL_WEEKDAYS]
}

function normalizeRanges(value: unknown): SenderActiveScheduleRange[] {
  if (!Array.isArray(value)) return []
  return value.flatMap((item) => {
    if (!isPlainObject(item)) return []
    const start = typeof item.start === 'string' ? item.start.trim() : ''
    const end = typeof item.end === 'string' ? item.end.trim() : ''
    if (!TIME_PATTERN.test(start) || !TIME_PATTERN.test(end) || start === end) return []
    return [{ start, end }]
  })
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

File: runtime/src/main/java/io/github/magisk317/relay/domain/sender/SenderActiveSchedule.kt
package io.github.magisk317.relay.domain.sender

import io.github.magisk317.relay.common.constant.MessageType
import java.io.Serializable
import java.time.DayOfWeek
import java.time.LocalDateTime

data class SenderActiveSchedule(
    val sms: SenderActiveScheduleRule = SenderActiveScheduleRule(),
    val appNotify: SenderActiveScheduleRule = SenderActiveScheduleRule(),
    val callNotify: SenderActiveScheduleRule = SenderActiveScheduleRule(),
) : Serializable

data class SenderActiveScheduleRule(
    val enabled: Boolean = false,
    val mode: String = SenderActiveScheduleConst.MODE_BLACKLIST,
    val weekdays: List<Int> = SenderActiveScheduleConst.ALL_WEEKDAYS,
    val ranges: List<SenderActiveScheduleRange> = emptyList(),
) : Serializable

data class SenderActiveScheduleRange(
    val start: String = "",
    val end: String = "",
) : Serializable

data class SenderActiveScheduleSummary(
    val smsRanges: Int,
    val appNotifyRanges: Int,
    val callNotifyRanges: Int,
)

object SenderActiveScheduleConst {
    const val MODE_BLACKLIST = "blacklist"
    const val MODE_WHITELIST = "whitelist"
    val ALL_WEEKDAYS: List<Int> = (DayOfWeek.MONDAY.value..DayOfWeek.SUNDAY.value).toList()
    private val TIME_PATTERN = Regex("^([01]\\d|2[0-3]):([0-5]\\d)$")

    fun parseMinutes(time: String): Int? {
        val match = TIME_PATTERN.matchEntire(time.trim()) ?: return null
        val hour = match.groupValues[1].toIntOrNull() ?: return null
        val minute = match.groupValues[2].toIntOrNull() ?: return null
        return hour * 60 + minute
    }
}

object SenderActiveScheduleEvaluator {
    fun sanitize(schedule: SenderActiveSchedule?): SenderActiveSchedule {
        val safeSchedule = schedule ?: SenderActiveSchedule()
        return SenderActiveSchedule(
            sms = sanitizeRule(safeSchedule.sms),
            appNotify = sanitizeRule(safeSchedule.appNotify),
            callNotify = sanitizeRule(safeSchedule.callNotify),
        )
    }

    fun sanitizeRule(rule: SenderActiveScheduleRule?): SenderActiveScheduleRule {
        val safeRule = rule ?: SenderActiveScheduleRule()
        val mode = if (safeRule.mode == SenderActiveScheduleConst.MODE_WHITELIST) {
            SenderActiveScheduleConst.MODE_WHITELIST
        } else {
            SenderActiveScheduleConst.MODE_BLACKLIST
        }
        val weekdays = safeRule.weekdays
            .filter { it in SenderActiveScheduleConst.ALL_WEEKDAYS }
            .distinct()
            .ifEmpty { SenderActiveScheduleConst.ALL_WEEKDAYS }
        val ranges = safeRule.ranges.mapNotNull { range ->
            val start = range.start.trim()
            val end = range.end.trim()
            val startMinutes = SenderActiveScheduleConst.parseMinutes(start) ?: return@mapNotNull null
            val endMinutes = SenderActiveScheduleConst.parseMinutes(end) ?: return@mapNotNull null
            if (startMinutes == endMinutes) return@mapNotNull null
            SenderActiveScheduleRange(start = start, end = end)
        }
        return SenderActiveScheduleRule(
            enabled = safeRule.enabled && ranges.isNotEmpty(),
            mode = mode,
            weekdays = weekdays,
            ranges = ranges,
        )
    }

    fun summarize(schedule: SenderActiveSchedule?): SenderActiveScheduleSummary {
        val safeSchedule = sanitize(schedule)
        return SenderActiveScheduleSummary(
            smsRanges = countEnabledRanges(safeSchedule.sms),
            appNotifyRanges = countEnabledRanges(safeSchedule.appNotify),
            callNotifyRanges = countEnabledRanges(safeSchedule.callNotify),
        )
    }

    fun isAllowed(
        schedule: SenderActiveSchedule?,
        messageType: MessageType,
        now: LocalDateTime = LocalDateTime.now(),
    ): Boolean {
        val safeSchedule = sanitize(schedule)
        val rule = when (messageType) {
            MessageType.SMS_CODE,
            MessageType.SMS_PLAIN,
            -> safeSchedule.sms

            MessageType.APP_NOTIFY -> safeSchedule.appNotify
            MessageType.CALL_NOTIFY -> safeSchedule.callNotify
        }
        return isRuleAllowed(rule, now)
    }

    fun isRuleAllowed(
        rule: SenderActiveScheduleRule,
        now: LocalDateTime = LocalDateTime.now(),
    ): Boolean {
        if (!rule.enabled) return true
        val matched = matchesAnyRange(rule, now)
        return when (rule.mode) {
            SenderActiveScheduleConst.MODE_WHITELIST -> matched
            else -> !matched
        }
    }

    private fun matchesAnyRange(rule: SenderActiveScheduleRule, now: LocalDateTime): Boolean {
        val weekday = now.dayOfWeek.value
        val previousWeekday = if (weekday == DayOfWeek.MONDAY.value) {
            DayOfWeek.SUNDAY.value
        } else {
            weekday - 1
        }
        val currentMinutes = now.hour * 60 + now.minute
        return rule.ranges.any { range ->
            val startMinutes = SenderActiveScheduleConst.parseMinutes(range.start) ?: return@any false
            val endMinutes = SenderActiveScheduleConst.parseMinutes(range.end) ?: return@any false
            when {
                startMinutes < endMinutes -> {
                    weekday in rule.weekdays && currentMinutes in startMinutes until endMinutes
                }
                else -> {
                    (weekday in rule.weekdays && currentMinutes >= startMinutes) ||
                        (previousWeekday in rule.weekdays && currentMinutes < endMinutes)
                }
            }
        }
    }

    private fun countEnabledRanges(rule: SenderActiveScheduleRule): Int {
        return if (rule.enabled) rule.ranges.size else 0
    }
}
